docs/authentication.md

```diff
@@ -0,0 +1,48 @@
+# Authentication
+
+## Authentication modes
+
+The API provide 3 authentication modes.
+- Session ID: used by the DRF UI (API browser).
+- Bearer token: used by the SDK (substra client).
+- JWT cookie: used by the substra frontend.
+
+### Session ID
+
+- URL: `/api-auth`
+- Implemented in `backend.urls` [module](../backend/backend/url.py)
+- Based on DRF `SessionAuthentication` [scheme](https://www.django-rest-framework.org/api-guide/authentication/#sessionauthentication)
+  * The session key is stored on the server side (in `django_session` table), alongside an expiration date and an encoded dict of data containing the user ID.
+  * The session key is a random string encoded in ASCII format. [Source](https://docs.djangoproject.com/fr/4.1/topics/http/sessions)
+  * The server send the session key to the client, via a cookie, which should be included in the requests.
+
+### Bearer token
+
+- URL: `/api-token-auth`
+- Implemented in `backend.views` [module](../backend/backend/views.py)
+- Based on DRF `TokenAuthentication` [scheme](https://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication)
+  * The token is stored on server side (in `authtoken_token` table), alongside a creation date and and the user ID.
+  * The token is a random string encoded in hexadecimal format. [Source](https://github.com/encode/django-rest-framework/blob/master/rest_framework/authtoken/models.py)
+  * The token should be included in the `Authorization` HTTP header
+- The API has a custom layer to handle the token expiration based on its creation date. [Source](../libs/expiry_token_authentication.py)
+
+### JWT cookie
+
+- URLs: `/me/login`, `/me/refresh`
+- Implemented in `users.views.authentication` [module](../backend/users/views/authentication.py)
+- Based on `rest_framework_simplejwt` [library](https://django-rest-framework-simplejwt.readthedocs.io/en/latest/index.html).
+  * Generate a [JWT token](https://jwt.io/introduction)
+    - Header: contains the signing algorithm, encoded in Base 64.
+    - Payload: contains authentication data (user ID, expiration date) encoded in Base 64.
+    - Signature: the header and the payload encrypted with a secret.
+- The API has a custom layer to split the JWT token in 3 cookies.
+  * A long lifespan cookie which contains the header and the payload.
+  * A short lifespan cookie which contains the signature, using HTTP only (not accessible by JS code to prevent from Cross-Site-Scripting / XSS attacks).
+  * A short lifespan cookie which contains the refreshing expiration date, using HTTP only.
+
+## JWT authentication workflow
+
+1. The client check for existing cookie
```

Comment:
1. The client makes a call to `/me/refresh`. 
  * If there was a valid refresh cookie, they get new header+payload and signature cookies.
  * If the call to `/me/refresh` fails (missing or expired refresh cookie), then the client sends login and password to `/me/login` to get all 3 cookies (refresh, header+payload, signature)
2. Client then makes authenticated calls by:
  * sending the signature cookie 
  * and including the contents of the header+payload cookie in a `Authorization: JWT {header+payload}` HTTP header.
4. If this call is rejected for auth reasons, the client tries `/me/refresh` to renew the 2 auth cookies and tries the call again.